### ISSUE TITLE
Define and measure time to value

### ISSUE THREAD

**Issue by malomarrec** [sourcegraph/sourcegraph#32664]
### Context

See batch [changes strategy page](https://handbook.sourcegraph.com/strategy-goals/strategy/code-graph/batch-changes/#themes)

### Problem to solve

Define and implement a metric that is a good enough proxy for time to value and start measuring it.

**Comment by malomarrec** [comment]
Thought dump here: a metric like this will always be imperfect, but let me take a stab at defining this further:

When a new user that has not created any batch change yet comes in, we measure the time elapsed between when he starts using batch changes and when he succeeds / gets value.
- "starts using batch changes" is defined as when a user first previews a batch change
- "gets value" is defined as when that user applies the batch change

We should cohort this, and start the cohort on the date the users first previews a batch change. 
So in other words: for all users that ran their first ever batch changes preview, how many have applied it? how long did it take on average.

cohort month| current month | apply rate | time to apply
---|----|---|----|
2021-01 |2021-01 | 20% | 14 days
2021-01 |2021-02 | 40% | 18 days
2021-02 |2021-02 | 18% | 12 days

**Comment by danielmarquespt** [comment]
Are you thinking of linking the events to the same batch change? Or interpreting the events even if they are not part of the same batch change?

To be clear: Are you proposing tracking the time it takes from first previewing a batch change, to the first time _any_ batch change is applied?  Or tracking the time between previewing a batch change and applying _that same_ batch change?

**Comment by malomarrec** [comment]
The second one makes more sense actually @danielmarquespt

**Comment by danielmarquespt** [comment]
I think both options have their place, but I think they bring different nuances:

Measuring the time it takes for the user to complete a single batch change gives us a measure of how easy/quick it is to use. (candidate 1)

while

I think measuring the time it takes between the user being _aware_ of the feature (Aka previewing something) and actually using it, gives us a measure of the value of the feature.  (candidate 2)

Sure we are going to catch all the scenarios where the user was just playing around with the feature to see how it works and never applies a change. But I think these are the scenarios we want to see happening and understand why they are happening. At least while we are looking for traction.

**Comment by danielmarquespt** [comment]
(I think this need to be said: I still have to try doing a Batch Change myself, so everything I'm commenting so far is based on a lot of assumptions of how the flow works 😅)

**Comment by malomarrec** [comment]
It's great to have a fresh look :)

**Comment by chrispine** [comment]
Thanks for this, Malo!

**Comment by malomarrec** [comment]
One comment from @eseliger 
> Would time-to-value measure the time span from someone engages with batch changes or from the user is fully set up for batch changes and can start creating changesets? Wondering if the onboarding burden should be part of that measurement.

What would be a good measure of that? I'd love to hear ideas on how we can measure the entire experience end to end: at what event do you start, and at what event do you end?

**Comment by adeola-ak** [comment]
As someone who played around alot with batch changes without applying a change for a while, I could see how we'd get alot of that data that may skew what we are trying to measure. Do we have a general idea of how our customers engineers get made aware of BC after the feature is purchased? Assuming there are a few calls with a few engineers and then rest are expected discover and learn 'on their own' maybe we could build some type of tutorial mode which will be outside of these metrics to get more accurate data? When I was learning nextJS, there was some type of a tutorial mode which made me really eager to learn and play around with it

**Comment by malomarrec** [comment]
In light of more recent conversations, and our [q3 strat update](https://github.com/sourcegraph/handbook/pull/4168), I think we need to measure two things:

- iteration time: that's, roughly, the time between "I start a run" and "I see a preview". We can measure it only on SSBC to make it more feasible.
- time to value: the time between "some decides to use batch changes" and "someone gets value". We can, again, assume it's only SSBC to avoid complexity, and define it as the time between "someone starts using batch changes" (eg. starts playing with the spec editor) and the time of the first apply.

Our analytics infrastructure is changing, and it sounds like we can now have events/session data so I'm gonna work with @ebrodymoore to understand what we could do here.

**Comment by malomarrec** [comment]
In fact, all of this depends on our ability to [collect event data](https://docs.google.com/document/d/1f5NDBrk-qNZsgnAQNPRfDShGm3SyRkT38HxMZeYaJAE/edit)
